IntegrationTests/UITests/UITest.cs

```diff
@@ -0,0 +1,231 @@
+﻿using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using CommonTestUtils;
+using GitUI;
+using NUnit.Framework;
+
+namespace GitExtensions.UITests
+{
+    public static class UITest
+    {
+        public static void RunForm<T>(
+            Action showForm,
+            Func<T, Task> runTestAsync)
+            where T : Form
+        {
+            LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} entry");
+            var idleCompletionSource = new TaskCompletionSource<VoidResult>();
+
+            // Start runAsync before calling showForm.
+            // The latter might block until the form is closed, especially if using Application.Run(form).
+
+            // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
+            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
+            // complete.
+            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            {
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test entry");
+
+                // Wait for the form to be opened by the test thread.
+                await idleCompletionSource.Task;
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test idle");
+                var form = Application.OpenForms.OfType<T>().Single();
```

Comment:
❗️ This should not have changed from the original form.

```
await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
await WaitForIdleAsync();
```

Reply:
This is part of the major race condition. It cannot work with a separate function `WaitForIdleAsync`.
Refer to 896be655c6b773a0157040b08bce4f0e7618dd7b.